app/controllers/api/v1/books/populate_controller.rb

```diff
@@ -12,46 +15,30 @@ class PopulateController < Api::ApiController
         # We get a messy set of params in this method, so we don't use the normal
         # approach of strong parameter validation.  We hardcode the only params
         # we care about.
-        # rubocop:disable Metrics/AbcSize
-        # rubocop:disable Metrics/MethodLength
         # rubocop:disable Layout/LineLength
         def update
-          # down the road we should be using ActiveRecord-import and first_or_initialize instead.
-          # See how snapshots are managed.
-
-          is_book_empty = @book.query_doc_pairs.empty?
-
-          params[:query_doc_pairs].each do |pair|
-            query_doc_pair = @book.query_doc_pairs.find_or_create_by query_text: pair[:query_text],
-                                                                     doc_id:     pair[:doc_id]
-            query_doc_pair.position = pair[:position]
-            query_doc_pair.document_fields = pair[:document_fields].to_json
-
-            query = @case.queries.find_by(query_text: query_doc_pair.query_text)
-
-            query_doc_pair.information_need = query.information_need
-            query_doc_pair.notes = query.notes
-            query_doc_pair.options = query.options
+          puts "[PopulateController] Request Size is #{number_to_human_size(query_doc_pairs_params.to_s.bytesize)}"
 
-            if pair[:rating]
-              rating = query.ratings.find_by(doc_id: query_doc_pair.doc_id)
+          serialized_data = Marshal.dump(query_doc_pairs_params)
 
-              # we are smart and just look up the correct user id from rating.user_id via the database, no API data needed.
-              judgement = query_doc_pair.judgements.find_or_create_by user_id: rating.user_id
-              judgement.rating = pair[:rating]
-              judgement.user = rating.user
-              judgement.save!
-            end
-
-            query_doc_pair.save!
-          end
-
-          Analytics::Tracker.track_query_doc_pairs_bulk_updated_event current_user, @book, is_book_empty
+          puts "[PopulateController] the size of the serialized data is #{number_to_human_size(serialized_data.bytesize)}"
+          compressed_data = Zlib::Deflate.deflate(serialized_data)
+          puts "[PopulateController] the size of the compressed data is #{number_to_human_size(compressed_data.bytesize)}"
+          @book.populate_file.attach(io: StringIO.new(compressed_data), filename: "book_populate_#{@book.id}.bin.zip",
```

Comment:
@epugh I am not versed in `ActiveJob`, or rails generally.

Storing compressed blobs in the database is probably find considering what I imagine are relatively small deployments of Quepid.

Side note: extra documentation of the API is awesome

For the hosted Quepid service, it is a bit more worrying.

My web experience is a lot of Django and Celery. I would typically save a large file to a blob store like S3, and then download in the async tasks.

Or, if i could guarantee a file was only required by one async task, or that all related tasks would be on the same machine, I would save to local disk.

Reply:
thanks for the feedback.   Yeah, generally with ActiveStorage you back it with s3..   I went with DB for now just to avoid adding "yet anohter quepid setup task" and that the blobs shouldn't last a long time.   the app.quepid.com on heroku means that the local disk doesn't work cause each web worker has it's own local disk :-(..   WE'll see!